File: src/main/java/tasks/Tinkoff/TinkoffAlgoWinter2024/taskE_parallelC.py
def construct(mult, substr):
    if len(str(mult)) + 2 + len(substr) > mult * len(substr):
        return mult * substr

    return f'{mult}({substr})'


s = input()

memo = {}  # memo[(start_ind, substr_len)] = (mult, substr)
for substr_len in range(1, len(s) + 1):
    for start_ind in range(len(s) - substr_len + 1):
        if substr_len == 1:
            memo[(start_ind, 1)] = 1, s[start_ind]
            continue

        mult, substr = 1, s[start_ind:start_ind + substr_len]
        for sep_index in range(start_ind, start_ind + substr_len - 1):
            left_len = sep_index - start_ind + 1
            right_len = substr_len - left_len

            left_mult, left_substr = memo[(start_ind, left_len)]
            right_mult, right_substr = memo[(sep_index + 1, right_len)]

            if left_substr == right_substr:
                mult, substr = left_mult + right_mult, left_substr
            else:
                curr_str_len = min(len(str(mult)) + 2 + len(substr), mult * len(substr))
                merged_len = len(construct(left_mult, left_substr)) + len(construct(right_mult, right_substr))
                if curr_str_len > merged_len:
                    mult, substr = 1, construct(left_mult, left_substr) + construct(right_mult, right_substr)

        memo[(start_ind, substr_len)] = (mult, substr)

res = construct(*memo[0, len(s)])
print(res)

File: src/main/java/tasks/Tinkoff/TinkoffAlgoWinter2024/taskE_parallelC.cpp
#include <iostream>
#include <unordered_map>
#include <string>

using namespace std;

struct pair_hash {
    inline size_t operator()(const pair<int,int> & v) const {
        return v.first*31+v.second;
    }
};

unordered_map<pair<int, int>, pair<int, string>, pair_hash> memo;

string repeat(int mult, const string & substr) {
    string repeatedString = "";
    for (int i = 0; i < mult; ++i) {
        repeatedString += substr;
    }
    return repeatedString;
 }
string construct(int mult, const string & substr) {
    if (to_string(mult).length() + 2 + substr.length() > mult * substr.length()) {
        return repeat(mult, substr);
    }
    return to_string(mult) + "(" + substr + ")";
}

pair<int, string> compress(const string & s, int left, int right) {
    if (memo.find({left, right}) != memo.end()) {
        return memo[{left, right}];
    }
    if (right == left) {
        return {1, string(1, s[left])};
    }
    int mult = 1;
    string substr = s;
    string built = s;
    for (int sep = left; sep < right; sep++) {
        pair<int, string> left_res = compress(s, left, sep);
        pair<int, string> right_res = compress(s, sep + 1, right);
        if (left_res.second == right_res.second) {
            mult = left_res.first + right_res.first;
            substr = left_res.second;
            built = construct(mult, substr);
        } else {
            string left_substr = construct(left_res.first, left_res.second);
            string right_substr = construct(right_res.first, right_res.second);
            if (built.length() > left_substr.length() + right_substr.length()) {
                mult = 1;
                substr = left_substr + right_substr;
                built = substr;
            }
        }
    }
    memo[{left, right}] = {mult, substr};
    return {mult, substr};
}

int main() {
    string s;
    cin >> s;
    pair<int, string> res = compress(s, 0, s.length() - 1);
    string result = construct(res.first, res.second);
    cout << result << endl;
    return 0;
}

